File: packages/core-math/src/coupling.ts
/**
 * Mental coupling matrix C: the linear operator that maps a participant's
 * intended hinge angle vector to the hinge configuration they actually
 * picture. Diagonal entries encode self-fidelity (always 1 in the linear
 * regime); off-diagonal entries encode involuntary co-activation.
 *
 * The Kinematic Mental Mirroring (KMM) hypothesis predicts that C is
 * V_4-equivariant: reflecting the strip leaves C unchanged, and so does a
 * paper-flip. By Schur's lemma C then lives in a 4-dimensional parameter
 * family for the V_4 representation that arises on R^{N-1}; the simplest
 * non-trivial point in that family is the mirror coupling matrix below.
 *
 * Linear-algebra primitives are kept dependency-free: small matrix sizes
 * (n <= 50 hinges in v0.1) make a tight, audited implementation faster
 * than dispatching to a black-box numerical library.
 */

import {
  type GroupElement,
  GROUP_ELEMENTS,
  groupAction,
  reynoldsProject,
  type SymmetryGroup,
  equivarianceResidual,
} from "./symmetry.js";

export type Matrix = readonly (readonly number[])[];

export interface MentalCoupling {
  /** N x N coupling matrix where N is the hinge count. */
  readonly matrix: Matrix;
  readonly nHinges: number;
  /** Group used to constrain estimation; undefined if no symmetry was imposed. */
  readonly group?: SymmetryGroup;
  /** Provenance: closed form vs estimated from data. */
  readonly source: "analytic" | "empirical";
  /** Strength parameter beta in [0, 1] when source="analytic". */
  readonly beta?: number;
}

/**
 * Identity coupling: each hinge moves only itself. The null hypothesis
 * baseline against which empirical couplings are compared.
 */
export function identityCoupling(nHinges: number): MentalCoupling {
  if (!Number.isInteger(nHinges) || nHinges < 1) {
    throw new Error("identityCoupling: nHinges must be a positive integer");
  }
  const M: number[][] = Array.from({ length: nHinges }, (_, i) =>
    Array.from({ length: nHinges }, (_, j) => (i === j ? 1 : 0)),
  );
  return { matrix: M, nHinges, source: "analytic", beta: 0 };
}

/**
 * Mirror coupling matrix C(beta):
 *   C[i][i]      = 1
 *   C[i][n-1-i]  = beta if i != n-1-i
 *   else         = 0
 *
 * Eigenvalues:  {1+beta} (multiplicity floor(n/2)),
 *               {1-beta} (multiplicity floor(n/2)),
 *               {1}       (only when n is odd; central hinge is fixed by sigma).
 *
 * For beta in [0, 1] the matrix is contracting (all eigenvalues in [1-beta, 1+beta] >= 0)
 * and V_4-equivariant by construction.
 */
export function mirrorCouplingMatrix(
  nHinges: number,
  beta: number,
): MentalCoupling {
  if (!Number.isInteger(nHinges) || nHinges < 1) {
    throw new Error("mirrorCouplingMatrix: nHinges must be a positive integer");
  }
  if (!(beta >= 0) || !(beta <= 1)) {
    throw new Error(`mirrorCouplingMatrix: beta must be in [0, 1] (got ${beta})`);
  }
  const M: number[][] = Array.from({ length: nHinges }, () =>
    new Array<number>(nHinges).fill(0),
  );
  for (let i = 0; i < nHinges; i++) {
    M[i]![i] = 1;
    const j = nHinges - 1 - i;
    if (j !== i) {
      M[i]![j] = beta;
    }
  }
  return {
    matrix: M,
    nHinges,
    source: "analytic",
    beta,
  };
}

/**
 * Apply a coupling matrix to an intent vector: theta_mental = C theta_intent.
 * Output is clipped to [-angleMax, angleMax] when angleMax is provided.
 */
export function applyCoupling(
  coupling: MentalCoupling,
  thetaIntent: readonly number[],
  angleMax?: number,
): readonly number[] {
  const n = coupling.nHinges;
  if (thetaIntent.length !== n) {
    throw new Error(
      `applyCoupling: intent length ${thetaIntent.length} mismatched with coupling dim ${n}`,
    );
  }
  const out = new Array<number>(n).fill(0);
  for (let i = 0; i < n; i++) {
    let acc = 0;
    const row = coupling.matrix[i] ?? [];
    for (let j = 0; j < n; j++) {
      acc += (row[j] ?? 0) * (thetaIntent[j] ?? 0);
    }
    out[i] = angleMax === undefined ? acc : Math.max(-angleMax, Math.min(angleMax, acc));
  }
  return out;
}

/**
 * Estimate a coupling matrix from a stack of (intent, response) pairs by
 * ordinary least squares with optional Tikhonov regularisation:
 *
 *     C_hat = (X^T X + lambda I)^{-1} X^T Y .
 *
 * If a SymmetryGroup is supplied, the estimate is post-projected onto the
 * G-equivariant subspace via the Reynolds operator (Section 3.5 of MATH.md).
 *
 * Solver: Gauss-Jordan elimination on the small (n_hinges x n_hinges) Gram
 * matrix. n_hinges <= 50 keeps this comfortably under a millisecond.
 */
export function estimateCoupling(
  intents: readonly (readonly number[])[],
  responses: readonly (readonly number[])[],
  options: {
    nHinges: number;
    lambda?: number;
    group?: SymmetryGroup;
  },
): MentalCoupling {
  const { nHinges, lambda = 0, group } = options;
  if (intents.length !== responses.length) {
    throw new Error("estimateCoupling: intent and response counts disagree");
  }
  if (intents.length === 0) {
    throw new Error("estimateCoupling: at least one trial is required");
  }
  const K = intents.length;
  for (let k = 0; k < K; k++) {
    if ((intents[k]?.length ?? 0) !== nHinges) {
      throw new Error(`estimateCoupling: intents[${k}] has wrong length`);
    }
    if ((responses[k]?.length ?? 0) !== nHinges) {
      throw new Error(`estimateCoupling: responses[${k}] has wrong length`);
    }
  }

  // Build X^T X (nHinges x nHinges) and X^T Y (nHinges x nHinges).
  const xtx: number[][] = Array.from({ length: nHinges }, () =>
    new Array<number>(nHinges).fill(0),
  );
  const xty: number[][] = Array.from({ length: nHinges }, () =>
    new Array<number>(nHinges).fill(0),
  );
  for (let k = 0; k < K; k++) {
    const x = intents[k]!;
    const y = responses[k]!;
    for (let i = 0; i < nHinges; i++) {
      for (let j = 0; j < nHinges; j++) {
        xtx[i]![j]! += (x[i] ?? 0) * (x[j] ?? 0);
        xty[i]![j]! += (x[i] ?? 0) * (y[j] ?? 0);
      }
    }
  }
  if (lambda > 0) {
    for (let i = 0; i < nHinges; i++) xtx[i]![i]! += lambda;
  }

  // Solve xtx * B = xty for B.
  const B = solveLinearSystem(xtx, xty);
  // C such that response = C intent => Y = X C^T => B = (X^T X)^{-1} X^T Y = C^T.
  // Therefore C = B^T.
  let C: number[][] = transposeMatrix(B);
  if (group) {
    C = reynoldsProject(group, C).map((r) => [...r]);
  }
  return {
    matrix: C,
    nHinges,
    source: "empirical",
    group,
  };
}

/**
 * Frobenius norm of (M . expected_invariance - expected_invariance . M).
 * Returns 0 (within epsilon) when M is exactly G-equivariant.
 */
export function couplingEquivarianceResidual(
  coupling: MentalCoupling,
  group: SymmetryGroup,
): number {
  return equivarianceResidual(group, coupling.matrix);
}

/**
 * Power-method spectral radius (largest |eigenvalue|). For symmetric or
 * near-symmetric coupling matrices this converges quickly; we cap iterations
 * to keep the call cheap for UI use.
 */
export function spectralRadius(
  M: Matrix,
  options: { maxIter?: number; tol?: number } = {},
): number {
  const { maxIter = 200, tol = 1e-10 } = options;
  const n = M.length;
  if (n === 0) return 0;
  let v: number[] = new Array(n).fill(1);
  v[0] = 1;
  let lambda = 0;
  for (let iter = 0; iter < maxIter; iter++) {
    const Mv = new Array<number>(n).fill(0);
    for (let i = 0; i < n; i++) {
      for (let j = 0; j < n; j++) {
        Mv[i]! += (M[i]?.[j] ?? 0) * (v[j] ?? 0);
      }
    }
    const norm = Math.hypot(...Mv);
    if (norm < 1e-300) return 0;
    const next = Mv.map((x) => x / norm);
    let delta = 0;
    for (let i = 0; i < n; i++) {
      const d = (next[i] ?? 0) - (v[i] ?? 0);
      delta += d * d;
    }
    v = next;
    const lambdaNext = norm;
    if (Math.abs(lambdaNext - lambda) < tol && Math.sqrt(delta) < tol) {
      return lambdaNext;
    }
    lambda = lambdaNext;
  }
  return lambda;
}

/**
 * Effective rank: number of singular values exceeding threshold * sigma_max.
 * Implemented via Jacobi eigendecomposition of M M^T (sufficient for n <= 50).
 */
export function effectiveRank(M: Matrix, threshold = 1e-6): number {
  const n = M.length;
  if (n === 0) return 0;
  const MMT: number[][] = Array.from({ length: n }, () =>
    new Array<number>(n).fill(0),
  );
  for (let i = 0; i < n; i++) {
    for (let j = 0; j < n; j++) {
      let acc = 0;
      for (let k = 0; k < n; k++) {
        acc += (M[i]?.[k] ?? 0) * (M[j]?.[k] ?? 0);
      }
      MMT[i]![j] = acc;
    }
  }
  const eig = jacobiEigenvalues(MMT);
  const sigmas = eig.map((e) => Math.sqrt(Math.max(0, e))).sort((a, b) => b - a);
  const sigmaMax = sigmas[0] ?? 0;
  if (sigmaMax === 0) return 0;
  return sigmas.filter((s) => s > threshold * sigmaMax).length;
}

// ---- internal linear algebra ----

function solveLinearSystem(A: number[][], B: number[][]): number[][] {
  // Solve A X = B for X. A is n x n, B is n x m. Gauss-Jordan on augmented.
  const n = A.length;
  const m = B[0]?.length ?? 0;
  const aug: number[][] = Array.from({ length: n }, (_, i) =>
    [...(A[i] ?? new Array(n).fill(0)), ...(B[i] ?? new Array(m).fill(0))],
  );

  for (let col = 0; col < n; col++) {
    let pivotRow = col;
    let pivotMag = Math.abs(aug[col]?.[col] ?? 0);
    for (let r = col + 1; r < n; r++) {
      const mag = Math.abs(aug[r]?.[col] ?? 0);
      if (mag > pivotMag) {
        pivotMag = mag;
        pivotRow = r;
      }
    }
    if (pivotMag < 1e-12) {
      throw new Error("solveLinearSystem: singular matrix");
    }
    if (pivotRow !== col) {
      const tmp = aug[col]!;
      aug[col] = aug[pivotRow]!;
      aug[pivotRow] = tmp;
    }
    const pivot = aug[col]![col]!;
    for (let j = 0; j < n + m; j++) aug[col]![j]! /= pivot;
    for (let r = 0; r < n; r++) {
      if (r === col) continue;
      const factor = aug[r]![col]!;
      if (factor === 0) continue;
      for (let j = 0; j < n + m; j++) {
        aug[r]![j]! -= factor * aug[col]![j]!;
      }
    }
  }

  const X: number[][] = Array.from({ length: n }, (_, i) =>
    Array.from({ length: m }, (_, j) => aug[i]?.[n + j] ?? 0),
  );
  return X;
}

function transposeMatrix(M: number[][]): number[][] {
  const r = M.length;
  const c = M[0]?.length ?? 0;
  const out: number[][] = Array.from({ length: c }, () => new Array(r).fill(0));
  for (let i = 0; i < r; i++) {
    for (let j = 0; j < c; j++) {
      out[j]![i] = M[i]?.[j] ?? 0;
    }
  }
  return out;
}

/**
 * Symmetric Jacobi eigenvalue solver. Converges quadratically near a
 * diagonal matrix; for our small (<=50) symmetric inputs it's a one-line
 * dependency-free option.
 */
function jacobiEigenvalues(
  Sin: readonly (readonly number[])[],
  maxIter = 200,
  tol = 1e-12,
): number[] {
  const n = Sin.length;
  const S: number[][] = Sin.map((row) => [...row]);
  for (let iter = 0; iter < maxIter; iter++) {
    let maxOff = 0;
    let p = 0;
    let q = 1;
    for (let i = 0; i < n; i++) {
      for (let j = i + 1; j < n; j++) {
        const v = Math.abs(S[i]?.[j] ?? 0);
        if (v > maxOff) {
          maxOff = v;
          p = i;
          q = j;
        }
      }
    }
    if (maxOff < tol) break;
    const Spp = S[p]?.[p] ?? 0;
    const Sqq = S[q]?.[q] ?? 0;
    const Spq = S[p]?.[q] ?? 0;
    const theta = (Sqq - Spp) / (2 * Spq || 1e-300);
    const t = Math.sign(theta) / (Math.abs(theta) + Math.sqrt(theta * theta + 1));
    const c = 1 / Math.sqrt(1 + t * t);
    const s = t * c;
    for (let i = 0; i < n; i++) {
      const Sip = S[i]?.[p] ?? 0;
      const Siq = S[i]?.[q] ?? 0;
      S[i]![p] = c * Sip - s * Siq;
      S[i]![q] = s * Sip + c * Siq;
    }
    for (let j = 0; j < n; j++) {
      const Spj = S[p]?.[j] ?? 0;
      const Sqj = S[q]?.[j] ?? 0;
      S[p]![j] = c * Spj - s * Sqj;
      S[q]![j] = s * Spj + c * Sqj;
    }
  }
  return Array.from({ length: n }, (_, i) => S[i]?.[i] ?? 0);
}

// Re-export for users who want to inspect the underlying group.
export { GROUP_ELEMENTS };
export type { GroupElement };

File: python/origami_lab/src/origami_lab/coupling.py
"""Mental coupling matrix estimation and manipulation.

The coupling matrix C maps intended hinge angle vector to the configuration
actually pictured. Diagonal entries encode self-fidelity (== 1 in the linear
regime); off-diagonal entries encode involuntary co-activation.

The Kinematic Mental Mirroring (KMM) hypothesis predicts that C is
V_4-equivariant. By Schur's lemma C lives in a 4-dimensional parameter family.

Numerically mirrors TypeScript ``@kinemind/core-math`` coupling.ts.
"""

from __future__ import annotations

import logging
from dataclasses import dataclass, field
from typing import Optional

import numpy as np
from numpy.typing import NDArray
from scipy import linalg as sp_linalg

from origami_lab.symmetry import (
    SymmetryGroup,
    equivariance_residual,
    reynolds_project,
)

logger = logging.getLogger(__name__)


@dataclass(frozen=True)
class MentalCoupling:
    """Container for a coupling matrix and its provenance.

    Attributes:
        matrix: (n_hinges, n_hinges) coupling matrix.
        n_hinges: Number of hinges.
        group: Optional symmetry group used during estimation.
        source: 'analytic' if built analytically, 'empirical' if estimated.
        beta: Strength parameter in [0,1] when source='analytic'.
        warnings: List of diagnostic warning strings collected during estimation.
            Mirrors the adaptive singularity detection in coupling.ts
            solveLinearSystem. Non-fatal conditions (near-singular Gram matrix
            with regularisation applied) are stored here rather than raised.
    """

    matrix: NDArray[np.float64]
    n_hinges: int
    group: Optional[SymmetryGroup] = None
    source: str = "analytic"
    beta: Optional[float] = None
    warnings: tuple[str, ...] = field(default_factory=tuple)

    def __post_init__(self) -> None:
        """Validate and sanitize inputs."""
        object.__setattr__(
            self, "matrix", np.asarray(self.matrix, dtype=np.float64)
        )
        if self.matrix.shape != (self.n_hinges, self.n_hinges):
            raise ValueError(
                f"MentalCoupling: matrix shape {self.matrix.shape} does not "
                f"match n_hinges={self.n_hinges}"
            )
        if self.source not in ("analytic", "empirical"):
            raise ValueError(
                f"MentalCoupling: source must be 'analytic' or 'empirical' "
                f"(got '{self.source}')"
            )


def identity_coupling(n_hinges: int) -> MentalCoupling:
    """Return the identity (null-hypothesis) coupling matrix.

    Args:
        n_hinges: Number of hinges (>= 1).

    Returns:
        MentalCoupling with identity matrix (beta=0).

    Raises:
        ValueError: if n_hinges is not a positive integer.
    """
    if not isinstance(n_hinges, int) or n_hinges < 1:
        raise ValueError("identity_coupling: n_hinges must be a positive integer")
    M = np.eye(n_hinges, dtype=np.float64)
    return MentalCoupling(matrix=M, n_hinges=n_hinges, source="analytic", beta=0.0)


def mirror_coupling_matrix(n_hinges: int, beta: float) -> NDArray[np.float64]:
    """Compute the mirror coupling matrix C(beta) as a plain ndarray.

    C[i][i]      = 1
    C[i][n-1-i]  = beta  (if i != n-1-i)
    else         = 0

    Eigenvalues: {1+beta} (floor(n/2) times), {1-beta} (floor(n/2) times),
    {1} once when n is odd.

    Args:
        n_hinges: Number of hinges (>= 1).
        beta: Mirror coupling strength in [0, 1].

    Returns:
        ndarray shape (n_hinges, n_hinges).

    Raises:
        ValueError: if inputs are invalid.
    """
    if not isinstance(n_hinges, int) or n_hinges < 1:
        raise ValueError("mirror_coupling_matrix: n_hinges must be a positive integer")
    if not (0.0 <= beta <= 1.0):
        raise ValueError(
            f"mirror_coupling_matrix: beta must be in [0, 1] (got {beta})"
        )
    M = np.zeros((n_hinges, n_hinges), dtype=np.float64)
    for i in range(n_hinges):
        M[i, i] = 1.0
        j = n_hinges - 1 - i
        if j != i:
            M[i, j] = beta
    return M


def mirror_coupling(n_hinges: int, beta: float) -> MentalCoupling:
    """Build a MentalCoupling object from the mirror coupling matrix.

    Args:
        n_hinges: Number of hinges (>= 1).
        beta: Mirror coupling strength in [0, 1].

    Returns:
        MentalCoupling with analytic mirror structure.
    """
    M = mirror_coupling_matrix(n_hinges, beta)
    return MentalCoupling(matrix=M, n_hinges=n_hinges, source="analytic", beta=beta)


def apply_coupling(
    coupling: MentalCoupling,
    theta_intent: NDArray[np.float64],
    angle_max: Optional[float] = None,
) -> NDArray[np.float64]:
    """Apply coupling matrix to an intent vector: theta_mental = C theta_intent.

    Args:
        coupling: MentalCoupling container.
        theta_intent: Intent angles shape (n_hinges,).
        angle_max: Optional clamp limit (applied element-wise).

    Returns:
        Mental angles shape (n_hinges,).

    Raises:
        ValueError: if theta_intent length mismatches coupling dimension.
    """
    x = np.asarray(theta_intent, dtype=np.float64)
    if x.shape != (coupling.n_hinges,):
        raise ValueError(
            f"apply_coupling: intent shape {x.shape} mismatched with "
            f"coupling dim {coupling.n_hinges}"
        )
    out = coupling.matrix @ x
    if angle_max is not None:
        out = np.clip(out, -angle_max, angle_max)
    return out


def _solve_gram(
    XtX: NDArray[np.float64],
    XtY: NDArray[np.float64],
    n_hinges: int,
    lambda_: float,
) -> tuple[NDArray[np.float64], list[str]]:
    """Solve the normal equations XtX B = XtY with adaptive singularity detection.

    Mirrors the adaptive ``singularThreshold`` logic in coupling.ts
    ``solveLinearSystem``.  When the effective condition number suggests near-
    singularity but regularisation has been applied, a warning is emitted
    instead of raising (matching the TS chained-warning pattern).

    Args:
        XtX: Gram matrix (n_hinges, n_hinges), may include lambda ridge term.
        XtY: Cross-product matrix (n_hinges, n_hinges).
        n_hinges: Dimension n.
        lambda_: Regularisation strength (>= 0).

    Returns:
        Tuple (B, warnings) where B = XtX^{-1} XtY and warnings is a list
        of diagnostic strings (empty when no near-singularity was detected).

    Raises:
        ValueError: if the Gram matrix is singular and lambda_ == 0.
    """
    warns: list[str] = []
    # Adaptive threshold: eps_mach * n * ||XtX||_F (same formula as TS).
    frob = float(np.linalg.norm(XtX, "fro"))
    singular_threshold = max(2.22e-16 * n_hinges * frob, 1e-300)
    eigvals = np.linalg.eigvalsh(XtX)
    min_eig = float(eigvals[0])
    if min_eig < singular_threshold:
        msg = (
            f"estimate_coupling: Gram matrix near-singular "
            f"(min_eigval={min_eig:.3e}, threshold={singular_threshold:.3e})"
        )
        if lambda_ > 0.0:
            warns.append(msg + f"; ridge lambda={lambda_:.3e} applied")
            logger.warning(warns[-1])
        else:
            raise ValueError(msg + "; add lambda_ > 0 for regularisation")

    try:
        B = sp_linalg.solve(XtX, XtY, assume_a="sym")
    except sp_linalg.LinAlgError as exc:
        raise ValueError("estimate_coupling: Gram matrix is singular") from exc

    return B, warns


def estimate_coupling(
    intents: NDArray[np.float64],
    responses: NDArray[np.float64],
    n_hinges: int,
    lambda_: float = 0.0,
    use_v4: bool = False,
    group: Optional[SymmetryGroup] = None,
) -> NDArray[np.float64]:
    """Estimate coupling matrix by ordinary least squares (Tikhonov optional).

    C_hat = argmin_C ||Y - X C^T||_F^2 + lambda ||C||_F^2

    where X is the (K, n_hinges) intent matrix and Y is the (K, n_hinges)
    response matrix. If group is provided (or use_v4=True), the estimate is
    post-projected onto the G-equivariant subspace via the Reynolds operator.

    Args:
        intents: Intent angles shape (K, n_hinges).
        responses: Response angles shape (K, n_hinges).
        n_hinges: Number of hinges.
        lambda_: Tikhonov regularisation coefficient (default 0).
        use_v4: If True and group is None, auto-construct V_4 group.
        group: Optional SymmetryGroup for equivariant projection.

    Returns:
        Estimated coupling matrix ndarray shape (n_hinges, n_hinges).

    Raises:
        ValueError: if input shapes are inconsistent or Gram matrix is singular.
    """
    from origami_lab.symmetry import klein_four_strip  # local to avoid cycle

    X = np.asarray(intents, dtype=np.float64)
    Y = np.asarray(responses, dtype=np.float64)

    if X.ndim != 2 or Y.ndim != 2:
        raise ValueError("estimate_coupling: intents and responses must be 2-D arrays")
    if X.shape != Y.shape:
        raise ValueError("estimate_coupling: intents and responses shapes disagree")
    K, n = X.shape
    if n != n_hinges:
        raise ValueError(
            f"estimate_coupling: column count {n} does not match n_hinges={n_hinges}"
        )
    if K == 0:
        raise ValueError("estimate_coupling: at least one trial is required")

    # Build Gram matrix X^T X (n x n) and cross-product X^T Y (n x n).
    XtX = X.T @ X
    XtY = X.T @ Y

    if lambda_ > 0.0:
        XtX = XtX + lambda_ * np.eye(n_hinges, dtype=np.float64)

    B, _warns = _solve_gram(XtX, XtY, n_hinges, lambda_)
    C = B.T

    # Optional symmetry projection.
    effective_group = group
    if use_v4 and effective_group is None:
        effective_group = klein_four_strip(n_hinges)
    if effective_group is not None:
        C = reynolds_project(effective_group, C)

    return C


@dataclass(frozen=True)
class CouplingWithCI:
    """Coupling matrix estimate with bootstrapped confidence intervals.

    Attributes:
        C_hat: Point estimate (n_hinges, n_hinges).
        lower_95: 2.5th percentile of bootstrap distribution (n_hinges, n_hinges).
        upper_95: 97.5th percentile of bootstrap distribution (n_hinges, n_hinges).
        bootstrap_n: Number of bootstrap replicates used.
        method: 'non-parametric' or 'analytic'.
    """

    C_hat: NDArray[np.float64]
    lower_95: NDArray[np.float64]
    upper_95: NDArray[np.float64]
    bootstrap_n: int
    method: str

    def __post_init__(self) -> None:
        """Validate field types."""
        if self.method not in ("non-parametric", "analytic"):
            raise ValueError(
                f"CouplingWithCI: method must be 'non-parametric' or 'analytic' "
                f"(got '{self.method}')"
            )
        if self.bootstrap_n < 1:
            raise ValueError(
                f"CouplingWithCI: bootstrap_n must be >= 1 (got {self.bootstrap_n})"
            )


def estimate_coupling_with_ci(
    intents: NDArray[np.float64],
    responses: NDArray[np.float64],
    n_hinges: int,
    *,
    lambda_: float = 0.0,
    bootstrap_n: int = 1000,
    seed: int = 0,
    method: str = "non-parametric",
) -> CouplingWithCI:
    """Estimate coupling matrix with 95% bootstrap confidence intervals.

    Extends :func:`estimate_coupling` with non-parametric case bootstrap CIs.
    The point estimate ``C_hat`` is identical to ``estimate_coupling``'s output.

    Bootstrap procedure (non-parametric):
        For each replicate, resample K rows (with replacement) from the
        (intents, responses) paired dataset, refit via OLS, and collect
        the empirical 2.5/97.5 percentile for each matrix element.

    Args:
        intents: Intent angles shape (K, n_hinges).
        responses: Response angles shape (K, n_hinges).
        n_hinges: Number of hinges.
        lambda_: Tikhonov regularisation applied in each bootstrap fit.
        bootstrap_n: Number of bootstrap replicates (default 1000).
        seed: Random seed for reproducibility (default 0).
        method: Bootstrap method; currently only 'non-parametric' is supported.

    Returns:
        CouplingWithCI with C_hat, lower_95, upper_95, bootstrap_n, method.

    Raises:
        ValueError: if inputs are inconsistent or method is unsupported.
    """
    if method != "non-parametric":
        raise ValueError(
            f"estimate_coupling_with_ci: only 'non-parametric' method is supported "
            f"(got '{method}')"
        )
    X = np.asarray(intents, dtype=np.float64)
    Y = np.asarray(responses, dtype=np.float64)
    if X.ndim != 2 or Y.ndim != 2:
        raise ValueError(
            "estimate_coupling_with_ci: intents and responses must be 2-D arrays"
        )
    if X.shape != Y.shape:
        raise ValueError(
            "estimate_coupling_with_ci: intents and responses shapes disagree"
        )
    K, n = X.shape
    if n != n_hinges:
        raise ValueError(
            f"estimate_coupling_with_ci: column count {n} does not match n_hinges={n_hinges}"
        )
    if K < 2:
        raise ValueError("estimate_coupling_with_ci: at least 2 trials are required for bootstrap")
    if bootstrap_n < 1:
        raise ValueError(f"estimate_coupling_with_ci: bootstrap_n must be >= 1 (got {bootstrap_n})")

    # Point estimate.
    C_hat = estimate_coupling(X, Y, n_hinges, lambda_=lambda_)

    # Non-parametric bootstrap.
    rng = np.random.default_rng(seed)
    boot_samples: list[NDArray[np.float64]] = []

    for _ in range(bootstrap_n):
        idx = rng.integers(0, K, size=K)
        Xb = X[idx]
        Yb = Y[idx]
        XtX_b = Xb.T @ Xb
        XtY_b = Xb.T @ Yb
        if lambda_ > 0.0:
            XtX_b = XtX_b + lambda_ * np.eye(n_hinges, dtype=np.float64)
        try:
            B_b, _ = _solve_gram(XtX_b, XtY_b, n_hinges, lambda_)
            boot_samples.append(B_b.T)
        except ValueError:
            # Near-singular bootstrap sample: skip without counting as failure.
            logger.debug("estimate_coupling_with_ci: singular bootstrap sample skipped")
            continue

    if len(boot_samples) == 0:
        raise ValueError(
            "estimate_coupling_with_ci: all bootstrap samples were singular; "
            "increase lambda_ or provide more data"
        )

    stack = np.stack(boot_samples, axis=0)  # (B, n, n)
    lower_95 = np.percentile(stack, 2.5, axis=0)
    upper_95 = np.percentile(stack, 97.5, axis=0)

    return CouplingWithCI(
        C_hat=C_hat,
        lower_95=lower_95,
        upper_95=upper_95,
        bootstrap_n=len(boot_samples),
        method=method,
    )


def coupling_equivariance_residual(
    coupling: MentalCoupling,
    group: SymmetryGroup,
) -> float:
    """Frobenius distance of the coupling matrix to the G-equivariant subspace.

    Args:
        coupling: MentalCoupling container.
        group: SymmetryGroup.

    Returns:
        ||C - reynolds_project(G, C)||_F (float).
    """
    return equivariance_residual(group, coupling.matrix)


def spectral_radius(
    M: NDArray[np.float64],
    max_iter: int = 200,
    tol: float = 1e-10,
) -> float:
    """Power-method spectral radius (largest absolute eigenvalue).

    Args:
        M: Square matrix shape (n, n).
        max_iter: Maximum power iterations (default 200).
        tol: Convergence tolerance (default 1e-10).

    Returns:
        Spectral radius (float).
    """
    M = np.asarray(M, dtype=np.float64)
    n = M.shape[0]
    if n == 0:
        return 0.0
    v = np.ones(n, dtype=np.float64)
    v[0] = 1.0
    lam = 0.0
    for _ in range(max_iter):
        Mv = M @ v
        norm = float(np.linalg.norm(Mv))
        if norm < 1e-300:
            return 0.0
        next_v = Mv / norm
        delta = float(np.linalg.norm(next_v - v))
        v = next_v
        lam_next = norm
        if abs(lam_next - lam) < tol and delta < tol:
            return lam_next
        lam = lam_next
    return lam


def effective_rank(M: NDArray[np.float64], threshold: float = 1e-6) -> int:
    """Number of singular values exceeding threshold * sigma_max.

    Args:
        M: Matrix shape (m, n).
        threshold: Relative threshold (default 1e-6).

    Returns:
        Effective rank (int).
    """
    M = np.asarray(M, dtype=np.float64)
    if M.size == 0:
        return 0
    sv = np.linalg.svd(M, compute_uv=False)
    sigma_max = sv[0] if sv.size > 0 else 0.0
    if sigma_max == 0.0:
        return 0
    return int(np.sum(sv > threshold * sigma_max))
